google-cloud-storage/src/test/java/com/google/cloud/storage/ITUnbufferedResumableUploadTest.java

```diff
@@ -0,0 +1,117 @@
+import com.google.cloud.storage.it.runner.registry.Generator;
+import com.google.cloud.storage.spi.v1.StorageRpc;
+import com.google.common.collect.ImmutableList;
+import java.io.IOException;
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+@RunWith(StorageITRunner.class)
+@CrossRun(
+    backends = {Backend.PROD},
+    transports = {Transport.HTTP, Transport.GRPC})
+@Parameterized(ObjectSizes.class)
+public final class ITUnbufferedResumableUploadTest {
+
+  @Inject public Storage storage;
+  @Inject public BucketInfo bucket;
+  @Inject public Generator generator;
+
+  @Parameter public int objectSize;
+
+  public static final class ObjectSizes implements ParametersProvider {
+
+    @Override
+    public ImmutableList<Integer> parameters() {
+      return ImmutableList.of(256 * 1024, 2 * 1024 * 1024);
+    }
+  }
+
+  @Test
+  @Exclude(transports = Transport.GRPC)
+  public void json()
+      throws IOException, ExecutionException, InterruptedException, TimeoutException {
+    BlobInfo blobInfo = BlobInfo.newBuilder(bucket, generator.randomObjectName()).build();
+    Opts<ObjectTargetOpt> opts = Opts.empty();
+    final Map<StorageRpc.Option, ?> optionsMap = opts.getRpcOptions();
+    BlobInfo.Builder builder = blobInfo.toBuilder().setMd5(null).setCrc32c(null);
+    BlobInfo updated = opts.blobInfoMapper().apply(builder).build();
+
+    StorageObject encode = Conversions.json().blobInfo().encode(updated);
+    HttpStorageOptions options = (HttpStorageOptions) storage.getOptions();
+    Supplier<String> uploadIdSupplier =
+        ResumableMedia.startUploadForBlobInfo(
+            options,
+            updated,
+            optionsMap,
+            StorageRetryStrategy.getUniformStorageRetryStrategy().getIdempotentHandler());
+    JsonResumableWrite jsonResumableWrite =
+        JsonResumableWrite.of(encode, optionsMap, uploadIdSupplier.get(), 0);
+
+    UnbufferedWritableByteChannelSession<StorageObject> session =
+        ResumableMedia.http()
+            .write()
+            .byteChannel(HttpClientContext.from(options.getStorageRpcV1()))
+            .resumable()
+            .unbuffered()
+            .setStartAsync(ApiFutures.immediateFuture(jsonResumableWrite))
+            .build();
+
+    int additional = 13;
+    long size = objectSize + additional;
+    ByteBuffer b = DataGenerator.base64Characters().genByteBuffer(size);
+
+    UnbufferedWritableByteChannel open = session.open();
+    int written = open.write(b);
+    assertThat(written).isEqualTo(objectSize);
+    assertThat(b.remaining()).isEqualTo(additional);
+
+    int writtenAndClose = open.writeAndClose(b);
+    assertThat(writtenAndClose).isEqualTo(additional);
```

Comment:
Can you add a test to validate the rewind step works as expected when unaligned PUTs are partially received to align with 256kib?

Reply:
This is validated by the assertion above `assertThat(b.remaining()).isEqualTo(additional);` the only way that assertion will succeed is by the rewind having taken place.